samples/nrf9160/serial_lte_modem/src/slm_at_icmp.c

```diff
@@ -0,0 +1,496 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+#include <logging/log.h>
+#include <zephyr.h>
+#include <stdio.h>
+#include <string.h>
+#include <net/socket.h>
+#include <nrf_socket.h>
+#include <modem_info.h>
+#include "slm_at_host.h"
+#include "slm_at_icmp.h"
+
+LOG_MODULE_REGISTER(icmp, CONFIG_SLM_LOG_LEVEL);
+
+#define INVALID_SOCKET		-1
+#define ICMP_MAX_URL		128
+
+#define MY_STACK_SIZE		KB(1)
+#define MY_PRIORITY		K_LOWEST_APPLICATION_THREAD_PRIO
+
+#define ICMP			0x01
+#define ICMP_ECHO_REQ		0x08
+#define ICMP_ECHO_REP		0x00
+#define IP_PROTOCOL_POS		0x09
+
+
+/*
+ * Known limitation in this version
+ * - IPv6 support
+ */
+
+/**@brief List of supported AT commands. */
+enum slm_icmp_at_cmd_type {
+	AT_ICMP_PING,
+	AT_ICMP_MAX
+};
+
+/**@ ICMP Ping command arguments */
+static struct ping_argv_t {
+	struct addrinfo *src;   /* Localhost public address */
+	struct addrinfo *dest;  /* Destination for the ping */
+	int len;
+	int waitms;
+	int count;
+	int interval;
+} ping_argv;
+
+/** forward declaration of cmd handlers **/
+static int handle_at_icmp_ping(enum at_cmd_type cmd_type);
+
+/**@brief SLM AT Command list type. */
+static slm_at_cmd_list_t m_icmp_at_list[AT_ICMP_MAX] = {
+	{AT_ICMP_PING, "AT#XPING", handle_at_icmp_ping},
+};
+
+static K_THREAD_STACK_DEFINE(my_stack_area, MY_STACK_SIZE);
+
+static at_cmd_handler_t m_callback;
+static char buf[64];
+
+static struct k_work_q my_work_q;
+static struct k_work my_work;
+
+/* global variable defined in different files */
+extern struct at_param_list m_param_list;
+extern struct modem_param_info modem_param;
+
+/******************************************************************
+ * AT#XPING adopted from AT#NPING implemented by R&D. Changes:
+ * - Support IPv4 only
+ * - Use nRF socket API instead of BSD API
+ * - Use ModemInfo to get local IP address
+ * - Use no Posix APIs
+ ******************************************************************/
+/* Taken from ipv4.c */
```

Comment:
This is ominous, it is not necessarily permitted to take code from somewhere else due to license restrictions.

If the license of ipv4.c permits us to copy the code then we should remove this comment.

If the license of ipv4.c does not permit us to copy the code then we should re-write this code.

Reply:
The same applies elsewhere.

Reply:
I'll remove it

Reply:
This was not removed everywhere.

Reply:
Sorry to have missed that. Just made a quick commit from home PC.